File: coffee-system/src/main/java/com/jesperancinha/coffee/system/manager/ProcessorAbstract.java
package com.jesperancinha.coffee.system.manager;

import com.jesperancinha.coffee.system.queues.QueueAbstract;
import com.jesperancinha.coffee.system.concurrency.QueueCallable;
import org.apache.log4j.Logger;

import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.Future;
import java.util.concurrent.ThreadPoolExecutor;

/**
 * Created by joaofilipesabinoesperancinha on 05-05-16.
 */
public abstract class ProcessorAbstract {

    public static final String SCHEDULED_TASK_FAILED_TO_EXECUTE = "scheduled task faild to execute!";

    private static final Logger logger = Logger.getLogger(ProcessorAbstract.class);

    public abstract QueueAbstract getExecutorServiceQueue();

    public void waitForAllCalls(QueueCallable queueCallable) {
       queueCallable.waitForCalls();
    }


    public abstract String getExecutorName(Callable<Boolean> callable);

    public void runAllCalls(QueueCallable queueCallable) {
        queueCallable.getAllCallables().stream().forEach(
                booleanCallable -> {
                    final Map<String, ThreadPoolExecutor> executorServiceMap = getExecutorServiceQueue().getExecutorServiceMap();
                    final String executorName = getExecutorName(booleanCallable);
                    final ThreadPoolExecutor threadPoolExecutor = executorServiceMap.get(executorName);
                    final Future<Boolean> submitResult = threadPoolExecutor.submit(booleanCallable);
                    addSubmitResult(submitResult, queueCallable);
                } //
        );    }

    private void addSubmitResult(Future<Boolean> submitResult, QueueCallable queueCallable) {
        queueCallable.addSubmitResult(submitResult);
    }
}

File: coffee-system/src/main/kotlin/org/jesperancinha/coffee/system/manager/ProcessorAbstract.kt
package org.jesperancinha.coffee.system.manager

import org.jesperancinha.coffee.system.api.concurrency.QueueCallable
import org.jesperancinha.coffee.system.queues.Queue
import java.util.concurrent.Callable
import java.util.concurrent.Future
import java.util.function.Consumer

/**
 * Created by joaofilipesabinoesperancinha on 05-05-16.
 */
abstract class ProcessorAbstract {
    abstract val executorServiceQueue: Queue
    open fun waitForAllCalls(queueCallable: QueueCallable) {
        queueCallable.waitForCalls()
    }

    abstract fun getExecutorName(callable: Callable<Boolean>): String
    open fun runAllCalls(queueCallable: QueueCallable) {
        queueCallable.allCallables.forEach { booleanCallable ->
                val executorServiceMap = executorServiceQueue.executorServiceMap
                val executorName = getExecutorName(booleanCallable)
                val threadPoolExecutor = executorServiceMap[executorName]
                val submitResult = threadPoolExecutor!!.submit(booleanCallable)
                addSubmitResult(submitResult, queueCallable)
            }
    }

     private fun addSubmitResult(submitResult: Future<Boolean>, queueCallable: QueueCallable) {
        queueCallable.addSubmitResult(submitResult)
    }

    companion object {
        const val SCHEDULED_TASK_FAILED_TO_EXECUTE = "scheduled task faild to execute!"
    }
}